### ISSUE TITLE
Docker is running failed to get console mode for stdout: The handle is invalid. containers, you need to switch to Windows containers.

### ISSUE THREAD

**Issue by mortenseifert** [microsoft/navcontainerhelper#3326]
**Describe the issue**
When running Docker Desktop 4.27.0 / 4.27.1 creating a container failes with the follwoing error
Docker is running failed to get console mode for stdout: The handle is invalid. containers, you need to switch to Windows containers.

Downgrading to 4.26.1 and the container is created - same script

**Scripts used to create container and cause the issue**
```
$artifactUrl = Get-BCArtifactUrl -type Sandbox -version 22 -country dk -select Latest

$secpasswd = ConvertTo-SecureString "password" -AsPlainText -Force
$credential = New-Object System.Management.Automation.PSCredential ("admin", $secpasswd)
New-BcContainer -accept_eula:$accept_eula `
  -accept_outdated:$true `
  -containername $containername `
  -auth "NavUserPassword" `
  -Credential $credential `
  -alwaysPull `
  -doNotExportObjectsToText `
  -usessl:$false `
  -updateHosts `
  -assignPremiumPlan `
  -artifactUrl $artifactUrl `
  -shortcuts None `
  -includeAL `
  -memoryLimit 8G
```

**Full output of scripts**
```
BcContainerHelper version 6.0.5
BC.HelperFunctions emits usage statistics telemetry to Microsoft
BcContainerHelper is version 6.0.5
BcContainerHelper is not running as administrator
UsePsSession is True
Host is Microsoft Windows 11 Enterprise - 10.0.22631.3085
New-BcContainer Telemetry Correlation Id: 3d621703-ce24-44d1-b688-0836f3090bb8
Docker is running failed to get console mode for stdout: The handle is invalid. containers, you need to switch to Windo
ws containers.
At C:\Program Files\WindowsPowerShell\Modules\BcContainerHelper\6.0.5\ContainerHandling\New-NavContainer.ps1:513 char:9
+         throw "Docker is running $dockerOS containers, you need to sw ...
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (Docker is runni...ows containers.:String) [], RuntimeException
    + FullyQualifiedErrorId : Docker is running failed to get console mode for stdout: The handle is invalid. containe
   rs, you need to switch to Windows containers.
```

**Additional context**
- does it happen all the time?
Yes, it happens every time.

- did it use to work?
When downgrading to Docker Desktop 4.26.1 the container is again created when running the script.

**Comment by TrippyZ** [comment]
I have a small bit of information to add as I have the same problem.  It does not occur if I user PowerShell ISE.  It only occurs if I use vscode-terminal or PowerShell.

**Comment by scsloan** [comment]
Same Issue with the default script from Microsoft's doc.

Docker Desktop on Windows 10 : 4.27.1
Powershell : 7.4.1

Script used:

```
$artifactUrl = Get-BcArtifactUrl -type sandbox -country us -select Latest
New-BCContainer -accept_eula -containerName mysandbox -artifactUrl $artifactUrl
```

Output: 

```

PowerShell credential request
Using Windows Authentication. Please enter your Windows credentials.
Password for user _______: **********************

BcContainerHelper is version 6.0.4
BcContainerHelper is running as administrator
HyperV is Enabled
UsePsSession is True
Host is Microsoft Windows 10 Enterprise - 10.0.19045.3930
New-BCContainer Telemetry Correlation Id: 823722da-4295-4b54-aec3-4226f9df8724
Exception: C:\Users\ss_admin\Documents\PowerShell\Modules\BcContainerHelper\6.0.4\ContainerHandling\New-NavContainer.ps1:513
Line |
 513 |          throw "Docker is running $dockerOS containers, you need to sw …
     |          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Docker is running failed to get console mode for stdout: The handle is invalid. containers, you need to switch
     | to Windows containers.
```

**Comment by mortenseifert** [comment]
I can confirm that running the script from PowerShell ISE does indeed allow me to create the container.

But if I check the PowerShell version, it is reported as the same in Windows PowerShell (Terminal) and Windows PoserShell ISE

PS C:\Users\mse> $PSVersionTable.PSVersion

Major  Minor  Build  Revision
-----  -----  -----  --------
5      1      22621  2506

**Comment by aholstrup1** [comment]
@mortenseifert I also bumped into the same problem earlier today 😞 I think the root cause here is a bug in v. 4.27.0 and 4.27.1 of docker (https://github.com/docker/for-win/issues/13891)

As per the issue, they should ship a fix for this soon but until then the best mitigation seems to be to avoid v. 4.27.0 and 4.27.1 of Docker. If you're already on of those versions you can try this workaround (https://github.com/docker/for-win/issues/13891#issuecomment-1924161940)

**Comment by mortenseifert** [comment]
@aholstrup1 thanks for looking into this further. I was link of thinking that perhaps it was not related to BcContainerHelper - but now the issue is here, for others to follow. 
I have downgraded my local Docker installation for now.

**Comment by ckuetbach** [comment]
4.27.2 still has this issue

**Comment by NavNab** [comment]
Same here.
Although the problem originates from Docker, the error message provided by BcContainerHelper can be confusing.

After debugging, the real error is: **failed to get console mode for stdout: The handle is invalid.**
The issue occurs on line 504 in the script **New-NavContainer**:
```powershell
$dockerVersion = docker version -f "{{.Server.Os}}/{{.Client.Version}}/{{.Server.Version}}"
```
Actually, any Docker command you execute returns an array. The first entry is always '**failed to get console mode for stdout: The handle is invalid.**', followed by the anticipated outcome of the command. For instance, running a simple **docker ps** won't visually reveal the problem. It's only when you copy the output to the clipboard that the issue becomes apparent.
```powershell
docker ps | clip
```
Refer to the screenshot provided below for a clearer understanding.
I tried to bypass the issue by altering the script:
```powershell
$dockerOS = $dockerVersion.Split('/')[1]
$dockerClientVersion = $dockerVersion.Split('/')[2]
$dockerServerVersion = $dockerVersion.Split('/')[3]
```
The code runs smoothly until another Docker command is called. Following the same approach, the code attempts to parse the output, leading to the first element ($somevariable[0]) holding '**failed to get console mode for stdout: The handle is invalid.**' This ends up causing the same problem again.
BcContainerHelper could benefit from some enhancements to improve its error handling, which would be greatly appreciated.

![image](https://github.com/microsoft/navcontainerhelper/assets/34076325/600ab8e5-ffca-49ab-8f71-f9b2bc7a8c60)

This workaround solves the issue for me. No need to downgrade your docker.
https://github.com/docker/for-win/issues/13891#issuecomment-1924161940
![image](https://github.com/microsoft/navcontainerhelper/assets/34076325/372b8ea7-d74c-490d-b26b-48266b0b8416)

**Comment by freddydk** [comment]
Just read that they will include a fix for this problem in docker 4.28
